### ISSUE TITLE
Persona 2 Innocent Sin's UI lines don't work at all resolutions

### ISSUE THREAD

**Issue by internetakias** [hrydgard/ppsspp#3332]
When BR's off the stripes that make up the backgrounds of certain UI elements are being spaced out correctly.
![screen00098](https://f.cloud.github.com/assets/5077063/1013465/05131db4-0b8a-11e3-9e37-b80893105a63.jpg)
When any form of BR is turned on however, the spacing gets messed up, causing certain stripes to be spaced further away from each other.
![screen00097](https://f.cloud.github.com/assets/5077063/1013462/f687425c-0b89-11e3-8cde-7d31961eba99.jpg)

**Comment by trostboot** [comment]
This is still present and behaves differently depending on the backend and rendering resolution. Win10, 64bit, v1.2.1-56-gcc4e7c4. AMD R9 270X, 16.2.1 driver.

OGL, 1xIR:
![ogl_1x_ulus10584_00008](https://cloud.githubusercontent.com/assets/17594477/13636809/2cedb96e-e604-11e5-8749-d0603f628e6c.png)
OGL, 2xIR:
![ogl_2x_ulus10584_00009](https://cloud.githubusercontent.com/assets/17594477/13636819/38b78e46-e604-11e5-9692-25bd6fa9d425.png)
D3D, 1xIR:
![d3d_1x_ulus10584_00010](https://cloud.githubusercontent.com/assets/17594477/13636830/4358c36a-e604-11e5-8e2e-ec3f3e49fce2.png)
D3D, 2xIR:
![d3d_2x_ulus10584_00011](https://cloud.githubusercontent.com/assets/17594477/13636838/505f6442-e604-11e5-8c41-c024cd387e7a.png)

Basically, it seems to be fine at 2x (and other multiples of 2) in OGL, and at 1x in D3D.

**Comment by Panderner** [comment]
Same happens in Eternal Punishment:
![Screenshot_2020-03-19-16-54-47-70_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37503397/77049162-8dd13b80-6a02-11ea-833c-d7edea04edbc.png)
![Screenshot_2020-03-19-16-55-33-24_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37503397/77049177-9295ef80-6a02-11ea-97ab-861aa8113a2e.png)

**Comment by hrydgard** [comment]
Yeah, this is an oldie but goodie. Will be fixed if we ever implement #8276 ...

**Comment by unknownbrackets** [comment]
I haven't tested this, but #15076 may help.

-[Unknown]

**Comment by Gamemulatorer** [comment]
Here's the results I don't get the real issue here though 😅

ppsspp v1.12.3-136

**Vulkan**
![Screenshot_2021-11-01-20-03-22-975_org ppsspp ppsspp](https://user-images.githubusercontent.com/37603562/139668691-88f9107a-3c01-413c-87bb-e8db28f480b3.jpg)

**PSP**
![Screenshot_2021-11-01-20-02-36-252_com vanced android youtube](https://user-images.githubusercontent.com/37603562/139668716-088df630-cebd-4e9a-97d7-62d1b8f5a84f.jpg)

**Comment by hrydgard** [comment]
Nice, looks pretty good!  

![image](https://user-images.githubusercontent.com/130929/139686469-9ab62d09-7669-43d5-9e0d-f0fffd70f95a.png)

Maybe not perfect, but not sure it's possible to get it perfect with the information we have from the game (just plain line draw commands).

**Comment by unknownbrackets** [comment]
I think that part always kinda looked that way (see even first screenshots), it was actually the cross-hatch shaded backgrounds this bug was about.

I'm going to reopen though, the lines need some work at some resolutions.

![Lines at 3x](https://user-images.githubusercontent.com/191233/139798776-268e2af1-2b8c-45b2-842c-ef81bfbe9151.png)

-[Unknown]

**Comment by hrydgard** [comment]
Ah I see, oh yeah, the hatched backgrounds are way better indeed.

**Comment by unknownbrackets** [comment]
Just to clarify status after that last change, here's a frame dump: [#3332_ULUS10584_persona2_lines.zip](https://github.com/hrydgard/ppsspp/files/7471469/3332_ULUS10584_persona2_lines.zip)

Correct (PSP):
![#3332_ULUS10584_persona2_lines](https://user-images.githubusercontent.com/191233/140237915-f0fb087a-9b7e-48d6-ae8d-427cf83c4907.png)

Software rendering:
![#3332_ULUS10584_persona2_lines_softgpu](https://user-images.githubusercontent.com/191233/140237932-fc759062-aef8-4fd7-b325-a162403e7763.png)

GLES/Vulkan (they're pretty much the same):
![#3332_ULUS10584_persona2_lines_vulkan](https://user-images.githubusercontent.com/191233/140237990-b26e7c7c-401d-4d4a-abed-f4346819e8a7.png)

As noted in #15091, the diagonal lines are still too think.

-[Unknown]

**Comment by hrydgard** [comment]
Thanks for reporting!

The missing gradient is likely due to some missing check in our rectangle drawing optimizations that have been added to the software renderer. I'll look into it soon, definitely before the 1.19 release (spring).

There's not really a good definition of line shapes that will work perfectly at higher resolutions when a game really was designed for a specific resolution, the lines interact with the bitmapped neighboring graphics. But it does seem like we could do better there too.

And yes, I know, #13079 is annoying... Will have to take a look at that again too.

**Comment by hrydgard** [comment]
a2abf9402bdb180731c99a419efcb67124aa6194 broke line gouraud shading with a logic mistake. fixing.

**Comment by hrydgard** [comment]
@Tayruu both renderers should now have the worst problems fixed, although some line shaping problems remain in the hardware renderer. Please test!

**Comment by Tayruu** [comment]
@hrydgard oh god I didn't know I got atted I'm sorry I didn't follow up for months uhhhh
I can confirm that on both Software and Hardware on v1.19.3 that gradients, outlines, and scanline texture effects look correct now. The only thing remaining is the odd misshaping with battle commands, but it's pretty minor and really only obvious on say, 4x render.

**Software render:**
<img width="448" height="128" alt="Image" src="https://github.com/user-attachments/assets/b9da9810-9f34-4422-931a-a2d4e3ff672f" />
**Hardware render, 1x:**
<img width="448" height="128" alt="Image" src="https://github.com/user-attachments/assets/10eec0c7-58af-4a2e-982b-3688916afcb3" />
**Hardware render, 4x:**
<img width="448" height="128" alt="Image" src="https://github.com/user-attachments/assets/c3976288-1001-490b-a8b8-eef2902f2509" />

**EDIT:** ... actually the more I look at it the more I get the feeling these assets are drawn with in-engine line drawing, as opposed to being texture assets. The reason why the edges look strange is because it's upscaling individual line draws. It's not an emulator problem, it's just the way they designed it makes more sense at a 1-pixel level.

**Comment by hrydgard** [comment]
Yup, that's exactly the problem. We could probably do some adjustment hacks to make these lines look good, but that might make similar lines in other games look worse. The PSP has the hardware capability to draw 1-pixel-wide lines, and knowing exactly what to do with those when rendering at higher resolutions really is content-dependent...

Thanks for reporting the new state!

**Comment by hrydgard** [comment]
Not much more to do here, closing.
